pkg/distributor/otel.go

```diff
@@ -335,27 +352,44 @@ func httpRetryableToOTLPRetryable(httpStatusCode int) int {
 	return httpStatusCode
 }
 
-// writeErrorToHTTPResponseBody converts the given error into a grpc status and marshals it into a byte slice, in order to be written to the response body.
-// See doc https://opentelemetry.io/docs/specs/otlp/#failures-1
-func writeErrorToHTTPResponseBody(reqCtx context.Context, w http.ResponseWriter, httpCode int, grpcCode codes.Code, msg string, logger log.Logger) {
+// writeErrorToHTTPResponseBody converts the given error into a gRPC status and marshals it into a byte slice, in order to be written to the response body.
+// See doc https://opentelemetry.io/docs/specs/otlp/#failures-1.
+func writeErrorToHTTPResponseBody(r *http.Request, w http.ResponseWriter, httpCode int, grpcCode codes.Code, msg string, logger log.Logger) {
 	validUTF8Msg := validUTF8Message(msg)
-	w.Header().Set("Content-Type", "application/octet-stream")
+	contentType := r.Header.Get("Content-Type")
+	if contentType == "" {
+		// Default to protobuf encoding.
+		contentType = "application/x-protobuf"
+	}
+	w.Header().Set("Content-Type", contentType)
 	w.Header().Set("X-Content-Type-Options", "nosniff")
-	if server.IsHandledByHttpgrpcServer(reqCtx) {
+	if server.IsHandledByHttpgrpcServer(r.Context()) {
 		w.Header().Set(server.ErrorMessageHeaderKey, validUTF8Msg) // If httpgrpc Server wants to convert this HTTP response into error, use this error message, instead of using response body.
 	}
 	w.WriteHeader(httpCode)
 
-	respBytes, err := proto.Marshal(status.New(grpcCode, validUTF8Msg).Proto())
-	if err != nil {
-		level.Error(logger).Log("msg", "otlp response marshal failed", "err", err)
-		writeResponseFailedBody, _ := proto.Marshal(status.New(codes.Internal, "failed to marshal OTLP response").Proto())
-		_, _ = w.Write(writeResponseFailedBody)
-		return
+	var respBytes []byte
+	var err error
+	st := status.New(grpcCode, validUTF8Msg).Proto()
+	if contentType == jsonContentType {
+		respBytes, err = json.Marshal(st)
+		if err != nil {
+			level.Error(logger).Log("msg", "otlp response marshal failed", "err", err)
+			writeResponseFailedBody, _ := json.Marshal(status.New(codes.Internal, "failed to marshal OTLP response").Proto())
+			_, _ = w.Write(writeResponseFailedBody)
+			return
+		}
+	} else {
+		respBytes, err = st.Marshal()
+		if err != nil {
+			level.Error(logger).Log("msg", "otlp response marshal failed", "err", err)
+			writeResponseFailedBody, _ := proto.Marshal(status.New(codes.Internal, "failed to marshal OTLP response").Proto())
+			_, _ = w.Write(writeResponseFailedBody)
+			return
+		}
 	}
```

Comment:
Could we remove some of this duplication by introducing the concept of a codec or something like that? Then we could have one codec for JSON and another for Protobuf, resolve the codec in one place, and then we wouldn't need so many conditions based on the content type here and above.

Reply:
Please check my refactoring.

Reply:
I think we could take this one step further: can we pass the response payload (eg. `st` here) to `writeOTLPResponse`, and then it can call `json.Marshal` or `proto.Marshal`? Then we shouldn't need to pass a function to `writeOTLPResponse`.

Reply:
I did some more refactoring, PTAL.